File: frontend/src/app/services/admin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({ providedIn: 'root' })
export class AdminService {
  private apiURL = 'http://localhost:8081/api/admins';

  constructor(private http: HttpClient) {}

  registerDriver(data: FormData): Observable<any> {
    return this.http.post(`${this.apiURL}/driver-register`, data);
  }

  getDriverRequests(): Observable<any> {
    return this.http.get(`${this.apiURL}/driver-requests`);
  }

  getUserById(id: number): Observable<any> {
    return this.http.get(`${this.apiURL}/users/${id}`);
  }

  getAllUsers(search?: string, sortBy?: string, sortDirection?: string, page?: number, size?: number): Observable<any> {
    const params: any = {};
    if (search) params.search = search;
    if (sortBy) params.sortBy = sortBy;
    if (sortDirection) params.sortDirection = sortDirection;
    if (page != null) params.page = String(page);
    if (size != null) params.size = String(size);
    return this.http.get(`${this.apiURL}/users`, { params });
  }

  reviewDriverRequest(requestId: number | string, dto: any): Observable<any> {
    return this.http.put(`${this.apiURL}/driver-requests/${requestId}`, dto);
  }

  blockUser(id: number, reason: string): Observable<any> {
    return this.http.put(`${this.apiURL}/block/${id}`, { reason });
  }

  unblockUser(id: number): Observable<any> {
    return this.http.put(`${this.apiURL}/unblock/${id}`, null);

  }

    getPriceConfigurations(): Observable<any> {
    return this.http.get(`${this.apiURL}/price-configs`);
  }

  updatePriceConfigurations(priceData: any): Observable<any> {
    return this.http.put(`${this.apiURL}/price-configs`, priceData);
  }

  getActiveRides(): Observable<any> {
    console.log('AdminService.getActiveRides() called');
    const activeRidesUrl = 'http://localhost:8081/api/rides/active-rides';
    console.log('GET request URL:', activeRidesUrl);
    return this.http.get(activeRidesUrl);
  }

  // Calls backend GET /report with query parameters.
  getReport(params: { startDate?: number | null; endDate?: number | null; drivers?: boolean; users?: boolean; personId?: string | number | null }): Observable<any> {
    const url = `${this.apiURL}/report`;
    const httpParams: any = {};
    if (params.startDate != null) httpParams.startDate = String(params.startDate);
    if (params.endDate != null) httpParams.endDate = String(params.endDate);
    if (params.drivers != null) httpParams.drivers = String(params.drivers);
    if (params.users != null) httpParams.users = String(params.users);
    if (params.personId != null) httpParams.personId = String(params.personId);
    console.log('AdminService.getReport() called, URL:', url, 'params:', httpParams);
    return this.http.get(url, { params: httpParams });
  }
}

File: mobile/app/src/main/java/com/example/ridenow/service/AdminService.java
package com.example.ridenow.service;

import com.example.ridenow.dto.admin.AdminChangesReviewRequestDTO;
import com.example.ridenow.dto.admin.AdminUserResponseDTO;
import com.example.ridenow.dto.admin.BlockUserRequestDTO;
import com.example.ridenow.dto.admin.AdminRideHistoryItemDTO;
import com.example.ridenow.dto.admin.DriverChangeRequestDTO;
import com.example.ridenow.dto.admin.PagedResponseDTO;
import com.example.ridenow.dto.admin.PriceConfigRequestDTO;
import com.example.ridenow.dto.admin.PriceConfigResponseDTO;
import com.example.ridenow.dto.report.ReportResponseDTO;
import com.example.ridenow.dto.user.UserItemDTO;
import com.example.ridenow.dto.user.UserResponseDTO;
import com.example.ridenow.dto.util.PageResponse;

import java.util.List;
import java.util.Map;

import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.GET;
import retrofit2.http.Multipart;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Part;
import retrofit2.http.PartMap;
import retrofit2.http.Path;
import retrofit2.http.Query;

public interface AdminService {

    @GET("/api/admins/driver-requests")
    Call<List<DriverChangeRequestDTO>> getDriverRequests();

    @PUT("/api/admins/driver-requests/{requestId}")
    Call<Void> reviewDriverRequest(@Path("requestId") long requestId, @Body AdminChangesReviewRequestDTO dto);

    @GET("/api/admins/users/{id}")
    Call<UserResponseDTO> getUserById(@Path("id") long id);

    @GET("/api/admins/price-configs")
    Call<PriceConfigResponseDTO> getPriceConfig();

    @PUT("/api/admins/price-configs")
    Call<Void> updatePriceConfig(@Body PriceConfigRequestDTO dto);

    @GET("/api/admins/users")
    Call<PagedResponseDTO<AdminUserResponseDTO>> getAllUsers(
            @Query("search") String search,
            @Query("sortBy") String sortBy,
            @Query("sortDirection") String sortDirection,
            @Query("page") int page,
            @Query("size") int size
    );

    @PUT("/api/admins/block/{id}")
    Call<Void> blockUser(@Path("id") Long id, @Body BlockUserRequestDTO dto);

    @PUT("/api/admins/unblock/{id}")
    Call<Void> unblockUser(@Path("id") Long id);

        @Multipart
        @POST("/api/admins/driver-register")
        Call<Map<String, Object>> registerDriver(@PartMap Map<String, RequestBody> partMap,
                                                                                         @Part MultipartBody.Part profileImage);

    @GET("/api/admins/report")
    Call<ReportResponseDTO> getReport(
            @Query("startDate") Long startDate,
            @Query("endDate") Long endDate,
            @Query("drivers") boolean drivers,
            @Query("users") boolean users,
            @Query("personId") String personId
    );
    @GET("/api/admins/all-users")
    Call<PageResponse<UserItemDTO>> getAllUsers(@Query("page") int page, @Query("size") int size,
                                                @Query("sortBy") String sortBy, @Query("sortDir") String sortDir);

    @GET("/api/admins/ride-history")
    Call<PageResponse<AdminRideHistoryItemDTO>> getRideHistory( @Query("id") long userId, @Query("page") int page,
                                                                @Query("size") int size, @Query("sortBy") String sortBy,
                                                                @Query("sortDir") String sortDir, @Query("date") Long date);
}
